File: src/DataScience/R/composite.R
file_node <- function(bytes) {
  list(kind = "file", bytes = bytes)
}

folder_node <- function(...) {
  list(kind = "folder", children = list(...))
}

node_size <- function(node) {
  if (node$kind == "file") {
    return(node$bytes)
  }
  sum(vapply(node$children, node_size, numeric(1)))
}

readme <- file_node(2)
docs <- folder_node(file_node(3), file_node(5))
root <- folder_node(readme, docs)

cat(sprintf("leaf=%d\n", node_size(readme)))
cat(sprintf("docs=%d\n", node_size(docs)))
cat(sprintf("root=%d\n", node_size(root)))

File: src/DataScience/Julia/composite.jl
abstract type Node end

struct FileLeaf <: Node
    bytes::Int
end

struct FolderComposite <: Node
    children::Vector{Node}
end

node_size(node::FileLeaf) = node.bytes
node_size(node::FolderComposite) = sum(node_size, node.children)

readme = FileLeaf(2)
docs = FolderComposite(Node[FileLeaf(3), FileLeaf(5)])
root = FolderComposite(Node[readme, docs])

println("leaf=$(node_size(readme))")
println("docs=$(node_size(docs))")
println("root=$(node_size(root))")
